### ISSUE TITLE
Perform tasks sequentially LESS and SASS before STYLES

### ISSUE THREAD

**Issue by garbinmarcelo** [JeffreyWay/laravel-mix#2961]
-   Laravel Mix Version: 6.0.18
-   Node Version (`node -v`): 14.16.1
-   NPM Version (`npm -v`): 6.14.12
-   OS: Windows 10

### Description:
I have a webpack file where I am compiling less, sass and grouping with styles. However, whenever I perform these tasks the webpack is not respecting the order, in this case my final .css file is not correct. Only after the 2 execution that ends up working out.

### Steps To Reproduce:
```javascript
const mix = require('laravel-mix');
const {CleanWebpackPlugin} = require('clean-webpack-plugin');
require('laravel-mix-merge-manifest');

mix.disableNotifications();

mix.options({
    processCssUrls: false,
    cleanCss: {
        level: {
            1: {specialComments: 'none'}
        }
    },
    postCss: [
        require('postcss-discard-comments')({removeAll: true})
    ],
    purifyCss: false,
    clearConsole: false
});

// webpack configuration
mix.webpackConfig({
    plugins: [
        new CleanWebpackPlugin({
            dry: false,
            cleanOnceBeforeBuildPatterns: [
                // not exclude files
                '!.htaccess',
                '!index.php',
                '!robots.txt',
                '!vendor',
                // exclude folder/files
                'static/css/site'
            ]
        })
    ]
});

// LESS
mix.less('resources/assets/less/skin-site.less', 'public/static/css/site');

// SASS
mix.sass('resources/assets/sass/site.scss', 'public/static/css/site');

// COMBINE CSS
mix.styles([
    'public/static/css/site/site.css',
    'public/static/css/site/skin-site.css',
    // Vendors
    'resources/assets/vendor/layerslider/6.9.2/css/layerslider.css',
], 'public/static/css/site/app-site.css');

if(mix.inProduction()){
    mix.version();
}

// merge mix-manifest.json
mix.mergeManifest();
```

**Comment by ToothlessRebel** [comment]
I'm seeing the same thing in Mix 6.0.22, Node 16.3.0, and npm 7.18.1.

The compiled Sass/Less files are being generated but it appears to be **after** the CSS is concatenated. 

`mix.styles()` will not throw an error if one of the provided files does not exist, so this completes "successfully".

I am chaining the methods `mix.less().sass().styles()` and tried calling them each separately but it not solve this issue.

**Comment by mikesaintsg** [comment]
Have you given the mix.after event hook a shot?
https://laravel-mix.com/docs/6.0/event-hooks
Let me know what you get. If it doesn't, I'll see if i can figure out a quick fix for you.

**Comment by garbinmarcelo** [comment]
> Have you given the mix.after event hook a shot?
> https://laravel-mix.com/docs/6.0/event-hooks
> Let me know what you get. If it doesn't, I'll see if i can figure out a quick fix for you.

I tried using before and then, but without success. I am currently using this package: https://github.com/kabbouchi/laravel-mix-merge-manifest#readme

What I did was split into 2 files, one for less and sass and the other just to combine the result of this first process. I solved the problem in a palliative way, but I'm still waiting for something better to solve it.

**Comment by mikesaintsg** [comment]
Thanks! Yeah, that seems like a good work around. I took a deeper look at mix and it only provides a single instance for a build, so a separate file is the only way to separate builds. Then there's also webpack that isn't built to run sequentially. Those are two big strikes, I don't see sequential builds coming to either packages any times soon.

In these rare cases, I would say gulp would be a better choice since they provide the ability to run builds sequentially using the series method.

https://gulpjs.com/docs/en/api/series

**Comment by garbinmarcelo** [comment]
A while ago I used gulp, I still have it in a legacy project, really it works. But I decided to set Laravel Mix as the default tool for assets, otherwise everything works fine.

Maybe this issue is a possible new feature :smile:

I think it's not just me who has this problem or is it? :flushed:

**Comment by andronocean** [comment]
Yep, same here. I've got a project that uses [copy-webpack-plugin](https://github.com/webpack-contrib/copy-webpack-plugin) to compile a bunch of JSON config files into a single object. I'm using `mix.combine()` to concatenate that with other files. In my case everything is working — but I would feel more confident about it if I could ensure that some things run sequentially.

I started using Mix because I wanted to get away from Gulp 😆

**Comment by thecrypticace** [comment]
I think it's entirely reasonable to expect .styles / .combine to run after the build but I'm not sure of the ramifications or requirements of making that work. I'll have to look into it. If we could wire it up directly as dependencies in webpack that would be preferable because it would then "just work" (heh famous last words)

**Comment by matthewjumpsoffbuildings** [comment]
I ran into this issue recently - i need to compile my pug templates to html first, before the SASS/Tailwind/PurgeCSS step, otherwise PurgeCSS strips _everything_ out of the output CSS file :(

**Comment by thecrypticace** [comment]
@matthewjumpsoffbuildings You should configure your purge configuration in tailwind to read from your pug templates to fix that. Priority selection is out of scope for a tool like Mix. At that point you really are looking for a tool like gulp because you need to perform arbitrary tasks in an arbitrary order.

**Comment by matthewjumpsoffbuildings** [comment]
unfortunately thats not an option, since the class names are generated dynamically in the pug templates, and so PurgeCSS _has_ to look at the output html files

so you are saying that having mix respect step definition sequence is just not an option? eg if your config is
```js
mix.pug({})
    .js({})
    .sass({})
```
there is no possibility of having that respected?

**Comment by thecrypticace** [comment]
1. Dynamic class name generation is explicitly recommended against in the tailwind documentation for this reason. It's fundamentally incompatible with tools like JIT.

2. The order of calls in the mix config for the most part does not matter and making that the case could (and very likely would) be a breaking change. It's also a non-trivial change because mix is built on top of webpack & dependency graphs rather than a sequence of explicit build steps. On top of this purge css sees what is _currently_ on the filesystem — not what has been built so far — so it would have to run after the build results had been written to disk. Which is effectively equivalent to running Mix twice.

3. The only possible workaround I can think of is to use `mix.before(async () => …)` to manually compile pug templates before mix runs the build.

**Comment by matthewjumpsoffbuildings** [comment]
yea i agree dynamic class name gen is inadvisable :) this is my biggest gripe with tailwind, the need for the whole purge step nonsense

for now i am getting around it with 2 separate mix files, which i call in sequence and use `laravel-mix-merge-manifest` 

i didnt know about the `.before()` method, will look into it. thanks

**Comment by thecrypticace** [comment]
There is one possible solution that may work in Mix v6.1 when multiple build support lands but I'm honestly not actually sure if it will. I'll keep it in mind and make a note if it does or does not.
